src/test/integ/highlight-grid.test.ts

```diff
@@ -0,0 +1,471 @@
+                    ],
+                    line: 0,
+                    col: 1,
+                },
+            ],
+        },
+        {
+            testName: "redrawing a virtual highlight over a double-wide character redraws both cells",
+            events: [
+                {
+                    line: 0,
+                    vimCol: 0,
+                    vimCells: [["a", 2], ["1"]],
+                    lineText: "你",
+                    tabSize: 4,
+                },
+                {
+                    line: 0,
+                    vimCol: 1,
+                    vimCells: [["2", 2]],
+                    lineText: "你",
+                    tabSize: 4,
+                },
+            ],
+            expectedRanges: [
+                {
+                    textType: "virtual" as const,
+                    highlights: [
+                        { hlId: 2, text: "a", virtText: "a" },
+                        { hlId: 2, text: "2", virtText: "2" },
+                    ],
+                    line: 0,
+                    col: 0,
+                },
+            ],
+        },
+        {
+            testName: "two highlights on the same line and same hlId should produce two ranges",
+            events: [
+                {
+                    line: 0,
+                    vimCol: 0,
+                    vimCells: [["h", 2], ["e"], ["l"], ["l"], ["o"]],
+                    lineText: "hello world hello",
+                    tabSize: 4,
+                },
+                {
+                    line: 0,
+                    vimCol: 12,
+                    vimCells: [["h", 2], ["e"], ["l"], ["l"], ["o"]],
+                    lineText: "hello world hello",
+                    tabSize: 4,
+                },
+            ],
+            expectedRanges: [
+                {
+                    textType: "normal" as const,
+                    hlId: 2,
+                    line: 0,
+                    startCol: 0,
+                    endCol: 5,
+                },
+                {
+                    textType: "normal" as const,
+                    hlId: 2,
+                    line: 0,
+                    startCol: 12,
+                    endCol: 17,
+                },
+            ],
+        },
+    ] as {
+        testName: string;
+        events: {
+            line: number;
+            vimCol: number;
+            vimCells: VimCell[];
+            lineText: string;
+            tabSize: number;
+        }[];
+        expectedRanges: HighlightRange[];
+    }[];
+
+    testCases.forEach(({ testName, events, expectedRanges }) => {
+        if (testName !== "allows overlaying virtual text on an existing line") return;
+        it(testName, () => {
+            const grid = new HighlightGrid(
+                1, // arbitrary
+                // fake the group store, only for testing
+                { normalizeHighlightId: (hlId) => hlId } as HighlightGroupStore,
+            ) as any as THighlightGrid;
```

Comment:
What is this cast for?

Reply:
No error when accessing private methods.

Reply:
Why do we need to access private methods in a test? This is a sign we should factor a code in the way that is more usable from our tests. I don't mean to suggest that unit tests are a panacea, or a perfect solution, but it is a method that I put in place primarily due to the difficulty of doing a full integration test. If we use them (which we do not have to, I just want a way to test highlights), we should ensure that we don't have to hack around our code in order to test it.